### ISSUE TITLE
Problem: It is impossible to implement a Kotlin like copy method in Dart if you have "nullable" variables

### ISSUE THREAD

**Issue by kasperpeulen** [dart-lang/language#137]
In Kotlin every data class is provided with a copy method. It copies the object and optionally it allows for setting new values. Here is an example:

```kotlin
import java.time.LocalDate

data class Todo(
    val body: String = "", 
    val completed: Boolean = false, 
    val dueDate: LocalDate? = null);

fun main() {
    val todo = Todo(body = "Do this", dueDate = LocalDate.now().plusDays(1));
    // user decide to remove the dueDate
    print(todo.copy(body = "Investigate this some day", dueDate = null));   
}
```

The copy method can be implemented like this in Kotlin:

```kotlin
    fun copy(
        body: String = this.body, 
        completed: Boolean = this.completed, 
        dueDate: LocalDate? = this.dueDate
    ): Todo {
        return Todo(body, completed, dueDate);
    }
```

However, doing the same in Dart doesn't work, as in Dart default values of parameters must be constant:

```dart
class Todo {
  final String body;
  final bool completed;
  final DateTime dueDate;

  Todo({this.body, this.completed, this.dueDate});

  Todo copy({
    String body = this.body,
    bool completed = this.completed,
    DateTime dueDate = this.dueDate,
  }) {
    return Todo(body: body, completed: completed, dueDate: dueDate);
  }
}
```

To allow for "default values" that are non-constant, often the following is recommended:

```dart
class Todo {
  final String body;
  final bool completed;
  final DateTime dueDate;

  Todo({this.body, this.completed, this.dueDate});

  Todo copy({
    String body,
    bool completed,
    DateTime dueDate,
  }) {
    body ??= this.body;
    completed ??= this.completed;
    dueDate ??= this.dueDate;
    return Todo(body: body, completed: completed, dueDate: dueDate);
  }
}
```

In many cases this works well. It's different in a subtle way, passing `null` has a special meaning now.  It means, "give me the default value, whatever that is". This works well for non-nullable variables, as null can now be used to mean something different. But also for nullable variables if the default value is also null. 

However, it makes one case impossible, if you have a "true" nullable variable with a non-null default value. In this case, `dueDate` is truely nullable. `null` means here that there exists no `dueDate` for this todo. However, this makes it impossible to use the copy method to remove the due date of the today:

```dart
main() {
  final todo = Todo(body: "Do this", dueDate: DateTime.now().add(Duration(days: 1)));
  print(todo.copy(body: "Investigate this some day", dueDate: null));
}
```

In summary: *It is impossible in Dart to give a nullable parameter a default non-constant value.*

I think the most obvious way to solve this would be to allow optional parameters to have a non-constant value. What is the reason for this restriction?

**Comment by eernstg** [comment]
We discussed allowing default values to be non-constant, e.g., allowing them to perform arbitrary computations in the enclosing scope (for an instance method that would be the instance scope of the enclosing class). Because of this, default values were _not_ included when we defined constant contexts, so we are in a sense preparing for that kind of generalization.

See https://github.com/dart-lang/sdk/issues/25572 for some earlier discussions on related topics.

**Comment by chriswiesner** [comment]
coming from Kotlin this is really cumbersome, i'd also appreciate to implement a simple copy mechanism

**Comment by munificent** [comment]
> What is the reason for this restriction?

I believe the original motivation is that it allows the compiler to check that an overriding method has the same default value as the method it overrides since the default values for both can be computed at compile time. Consider:

```dart
class A {
  int method([int n = 1]) => n;
}

void test(A a) {
  print(a.method());
}
```

A user who sees that code and looks up the declaration of `method()` on A probably assumes that since they are omitting the parameter, the value that it gets will be 1. But what they don't know is that their program is actually doing this:

```dart
class B extends A {
  int method([int n = 2]) => n;
}

main(List<String> args) {
  test(B());
}
```

The override in B has a different default. This is pretty confusing so Dart reports a warning if the default values don't match. It's not clear that this restriction actually carries its weight, though, and we have considered loosening this to allow non-const expressions there.

**Comment by tatumizer** [comment]
In parallel, we have a discussion  on the [same issue in sdk](https://github.com/dart-lang/sdk/issues/39897)

Specifically, it transpired that
> You can only pass arguments which match the type of the parameter. That means that you cannot pass null as an argument to the func written here because it needs an int. You can omit passing an argument, in which case the function will get the default value instead.

So, "null" doesn't have a special meaning "use default" (BTW, why do you think it does?). How to express that meaning is exactly the question under discussion in the cited thread.

**Comment by rrousselGit** [comment]
We actually can do a proper copyWith by using default values.

There are two tricks available:

- a combination of an abstract class + factory constructor, as showcased in a code-generator I released recently – see the generated file: https://github.com/rrousselGit/freezed/blob/master/example/lib/main.g.dart

- using `FutureOr`:

  ```dart
  class _ConstFuture<T> implements Future<T> {
    const _ConstFuture();
    @override
    dynamic noSuchMethod(Invocation invocation) {
      return super.noSuchMethod(invocation);
    }
  }

  class Example {
    int a;
    String b;

    Example copyWith({FutureOr<int> a = const _ConstFuture(), FutureOr<String> b = const _ConstFuture()}) {
      assert(a is int || a is _ConstFuture);
      assert(b is String || b is _ConstFuture);
      return Example()
        ..a = a is _ConstFuture ? this.a : a as int
        ..b = b is _ConstFuture ? this.b : b as String;
    }
  }
  ```

**Comment by tatumizer** [comment]
Very creative, but surreal. 
I was trying to put it in the same context where generated constructors reside. The rationale is simple: if we have a justification for generated constructors, the same justification applies to copy constructors like `copyWith`. For, in a general case, you need to list all the fields (which are many) again in your manually written copyWith method.

**Comment by rrousselGit** [comment]
Well, I may be biased but I feel that https://github.com/rrousselGit/freezed solves most problems in a very natural syntax.

The `FutureOr` variant is more of a work-around for all other situations that have a syntax different from `freezed`.

A `FutureOr` based `copyWith` could easily be generated as an extension method for example.

**Comment by tatumizer** [comment]
A purist will protest against FutureOr approach by arguing that the signature of the `copyWith` implies you can pass either a Future or a value, but an attempt to pass a Future will cause a runtime error.

IMO, the very fact that you have to be creative to trigger the use of default values adds extra weight (and urgency) to an argument in favor of a general typesafe solution, before these unconventional methods become popular :-)

**Comment by tatumizer** [comment]
There's another, more subtle, problem with the `copyWith` implementation: the use of the instance of private class as the default value. Suppose you want to create a copy of Example, but parameters "a" and "b" are obtained from some computation, and the values you receive are nullable (null stands for "useDefault"). A simple scenario: you have a map `<string, int?>{"a": 1, "b":null}` and you'd like to translate it into the call of `copyWith`:
```
var y  = x.copyWith(map["a"] ?? useDefault, map["b"] ?? useDefault);
```
But you can't really say `useDefault` - you don't know what to pass as a token for "useDefault", and even if you knew, you don't have access to the private class _ConstFuture.

**Comment by eernstg** [comment]
We have discussed the ability to access default values explicitly a number of times. One approach is to use null and let that stand for "please pass the default value" (but that clashes with other uses of null, that is, it doesn't work well for a parameter whose type is nullable). 

The approach that I usually recommend is to have explicit syntax. With suitable defaults (no pun intended, of course). We could then make `SomeClass.copyWith.default[0]` denote the default value of the first positional parameter of `copyWith` as declared in `SomeClass`; we could omit `SomeClass` and `copyWith` when it is used in an invocation of `copyWith` on a receiver whose static type is `SomeClass`, and we could omit `[0]` when it's used in the first positional argument. We would then have this:

```dart
var y  = x.copyWith(map["a"] ?? default, map["b"] ?? default);
```

**Comment by rrousselGit** [comment]
A better solution, in my opinion, would be the `if` expression:

```dart
var value = x.copyWith(x: if (condition) value)
```

Which would be equivalent to:

```dart
var value = condition
  ? x.copyWith(x: value)
  : x.copyWith();
```

**Comment by tatumizer** [comment]
@eernstg : 
passing null as a token for "default value" is, as you said, fraught with unintended consequences. Saying that those unintended consequences were accounted for won't be enough to pacify the critics of the language. But it's not only that. There are 2 other arguments against the idea: 1) it doesn't exactly fit into the concept of strong typing, and 2) it's a breaking change. Currently, null is not treated that way, so
```
  foo([int x=0])=>print(x); 
  foo(null);
```
prints null, not 0.

The idea of `SomeClass.copyWith.default[0]`, with natural shortcuts, is indeed less error-prone IMO. Still, a purist would point out that in the expression 
`var y  = x.foo(map["a"] ?? default, map["b"] ?? default);` 
the value substituted for "default" is based on the static type of `x`, which is not necessarily the same as the dynamic type of x, where the default value can be overridden. This kind of override would probably be quite rare in practice, but still...

**Comment by tatumizer** [comment]
@rrousselGit 
> `var value = x.copyWith(x: if (condition) value)`

Indeed, I was wondering why this was not done at the time when collection-if was introduced. I vaguely remember something about optional positional parameters - e.g. whether to shift the list or not if some parameter was omitted due to conditional if, but it's difficult to see it as a really serious argument. Or maybe the reason for not doing it was completely different :-)

**Comment by eernstg** [comment]
@tatumizer wrote:
>  "default" is based on the static type of x

Right, we have made it a warning (though not an error) to override such that a default value of a parameter changes, so that kind of override should indeed be rare. ;-)

> Also, it's not exactly clear .. index .. named parameters

We could use the name to access the default value of a named parameter: `default.a`.

**Comment by tatumizer** [comment]
@eernstg : Thanks! Yes, it makes sense now. :-)

**Comment by idkq** [comment]
Solution:

```
class Dog{
  final String name;
  final int age;
  Dog({this.name, this.age});
  
  Dog.fromMap(Map<String, dynamic> map): 
    name = map['name'] as String, age = map['age'] as int;
  Map<String, dynamic> toMap() => {'name': name, 'age': age};
  
  Dog copyWith(Map<String, dynamic> fields){
    var newObj = this.toMap();
    for (final field in fields.entries)
      newObj[field.key] = field.value;
    return Dog.fromMap(newObj);
  }
}

void main() {
  Dog buddy = Dog(name: 'Buddy', age: 12);
  Dog puppy = buddy.copyWith({'age': null});
  
  // OUTPUT=> Name: Buddy Age: null
  print('Name: ${puppy.name} Age: ${puppy.age}');
}


```

**Comment by spkersten** [comment]
@idkq With that solution you lose type-safety. That seems like a higher cost than opting out of null safety.

**Comment by idkq** [comment]
@spkersten You can make it null safe and its better than using futures, other classes, etc IMHO

**Comment by lrhn** [comment]
If you split your class into an interface and an implementation class, then you can have a default value that is not of the type of the original parameter:
```dart
abstract class NumBox {
  num? get value;
  NumBox(num? value) = _NumBox;
  NumBox replace({num? value});
}
class _NumBox implements NumBox {
  final num? value;
  _NumBox(this.value);
  NumBox replace({Object? value = "not a num"}) =>
    _NumBox((value is num?) ? value : this.value);
}
```
(not very useful with only the one property.

That's not always practical. That's where it would be nice (read: my current favorite idea in this area) to allow a parameter to have both an external type and a different *internal* type, with the default value making the conversion:
```dart
class NumBox {
  final num? value;
  NumBox(this.value);
  NumBox replace({num? value as Object? = "not a num"}) =>
    NumBox((value is num?) ? value : this.value);
}
```
Here the `replace` parameter `value` has type `num?` in the interface, but internally in the function it is widened `to `Object?`, and the default value can be taken from the wider type. (That would also allow `{int x as int? = null}` as a non-nullable value with a `null` default value when omitted).

It's something you can do if you split the class into an interface class and an implementation class, so it would be convenient if you could do that in one class declaration too, separating the *interface* signature from the *implementation* signature.

Alternatively, we can allow method forwarding like we do for factory constructors:
```dart
class NumBox {
  final num? value;
  NumBox(this.value);
  NumBox replace({num? value}) = _replace; // Forward call to `_replace`, missing arguments and all.
  NumBox _replace({Object? value = "not a num"}) =>
      NumBox((value is num?) ? value : this.value);
}
```

**Comment by jodinathan** [comment]
> A better solution, in my opinion, would be the `if` expression:
> 
> ```dart
> var value = x.copyWith(x: if (condition) value)
> ```
> 
> Which would be equivalent to:
> 
> ```dart
> var value = condition
>   ? x.copyWith(x: value)
>   : x.copyWith();
> ```

I also thought of this feature, however, how would this work with more than one parameter?  
ie

```dart
var value = x.copyWith(x: if (condition) value, y: if (otherCondition) otherValue)
```

**Comment by tatumizer** [comment]
If `condition` is false *for a specific parameter*, this parameter is copied from the object, as usual for omitted parameters.
There's no problem with named parameters. For positional parameters, there's a solution, too, but now that the discussion is moved into #2009, maybe we should discuss it there?
